### PR TITLE
[SYCL] Add support for scalar logical operators with group algorithms

### PR DESCRIPTION
**Pull request by jzc**

### CODE REVIEW

**Comment by jchlanda** [comment]
ping @JackAKirk

**Comment by JackAKirk** [comment]
> ping @JackAKirk

LGTM

**Review by elizabethandrews — COMMENTED** [review]
FE changes look ok to me naively. I am not familiar with this and hope runtime folks can review FE change as well.

**Review by bso-intel — APPROVED** [review]

File: clang/lib/Sema/SPIRVBuiltins.td [hunk 1]
```diff
@@ -952,3 +952,7 @@ foreach name = ["GroupUMin", "GroupUMax"] in {
 foreach name = ["GroupSMin", "GroupSMax"] in {
   def : SPVBuiltin<name, [ASIGenTypeN, UInt, UInt, ASIGenTypeN], Attr.Convergent>;
 }
+
+foreach name = ["GroupLogicalAndKHR", "GroupLogicalOrKHR"] in {
+  def : SPVBuiltin<name, [Bool, UInt, UInt, Bool], Attr.Convergent>;
```
Comment (AlexeySachkov) [inline clang/lib/Sema/SPIRVBuiltins.td:957]: According to [`SPV_KHR_uniform_group_instructions`](https://github.com/KhronosGroup/SPIRV-Registry/blob/main/extensions/KHR/SPV_KHR_uniform_group_instructions.asciidoc):

> 'Result Type' must be a scalar or vector of [Boolean type](https://github.com/KhronosGroup/SPIRV-Registry/blob/main/extensions/KHR/SPV_KHR_uniform_group_instructions.asciidoc#Boolean%20type).
> The type of 'X' must be the same as 'Result Type'.

I guess this change misses boolean vectors support. Note: I'm fine if that is outlined into a separate PR, but then we probably need to update the title to explicitly mention that only scalars are supported for now.

Comment (jzc) [reply clang/lib/Sema/SPIRVBuiltins.td:957]: Good catch, on the SYCL side, I think vector types should still work with the logical operators group algorithms, as the the current implementation of the group algorithms implement the vector group algorithms using the scalar group algorithms. But the SPIRV builtin will only be able to support a scalar value, so I can add a TODO comment mentioning that. I'll also add a test that uses vector types.

Comment (jzc) [reply clang/lib/Sema/SPIRVBuiltins.td:958]: There are a couple issues with the current implementation for group algorithms for vector types and the return type of `operator&&` for `vec<bool, N>`, so I will address those issues in another PR. I've renamed the PR to only mention support for scalar operations now.

Comment (AlexeySachkov) [reply clang/lib/Sema/SPIRVBuiltins.td:958]: > There are a couple issues with the current implementation for group algorithms for vector types and the return type of `operator&&` for `vec<bool, N>`, so I will address those issues in another PR. I've renamed the PR to only mention support for scalar operations now.

Works for me, thanks!

**Review by AlexeySachkov — APPROVED** [review]

**Review by elizabethandrews — APPROVED** [review]
FE changes LGTM but I'm not very familiar with the functionality. I trust @AlexeySachkov 's FE review, so approving.

**Pull request merged**